src/transport/grpc/convert.rs

```diff
@@ -0,0 +1,1347 @@
+                match d {
+                    $(
+                    AstarteType::$typ(val) => Ok(val.into()),
+                    )*
+                    AstarteType::Unset => Err(MessageHubProtoError::UnsetConversion),
+                }
+            }
+        }
+    }
+}
+
+impl_astarte_type_to_individual_data_conversion_traits!(
+    Double,
+    Integer,
+    Boolean,
+    LongInteger,
+    String,
+    BinaryBlob,
+    DateTime,
+    DoubleArray,
+    IntegerArray,
+    BooleanArray,
+    LongIntegerArray,
+    StringArray,
+    BinaryBlobArray,
+    DateTimeArray
+);
+
+impl TryFrom<astarte_message_hub_proto::AstarteMessage> for AstarteDeviceDataEvent {
+    type Error = MessageHubProtoError;
+
+    fn try_from(
+        astarte_message: astarte_message_hub_proto::AstarteMessage,
+    ) -> Result<Self, Self::Error> {
+        let astarte_sdk_aggregation = match optional_chain!(astarte_message.payload)? {
+            astarte_message_hub_proto::astarte_message::Payload::AstarteData(astarte_data_type) => {
+                match optional_chain!(astarte_data_type.data)? {
+                    astarte_message_hub_proto::astarte_data_type::Data::AstarteIndividual(
+                        astarte_individual,
+                    ) => Aggregation::Individual(
+                        optional_chain!(astarte_individual.individual_data)?.try_into()?,
+                    ),
+                    astarte_message_hub_proto::astarte_data_type::Data::AstarteObject(
+                        astarte_object,
+                    ) => {
+                        let astarte_sdk_aggregation =
+                            map_values_to_astarte_type(astarte_object.object_data)?;
+                        Aggregation::Object(astarte_sdk_aggregation)
+                    }
+                }
+            }
+            astarte_message_hub_proto::astarte_message::Payload::AstarteUnset(_) => {
+                Aggregation::Individual(AstarteType::Unset)
+            }
+        };
+
+        Ok(AstarteDeviceDataEvent {
+            interface: astarte_message.interface_name,
+            path: astarte_message.path,
+            data: astarte_sdk_aggregation,
+        })
+    }
+}
+
+impl TryFrom<astarte_message_hub_proto::types::InterfaceJson> for crate::Interface {
+    type Error = crate::Error;
+
+    fn try_from(
+        interface: astarte_message_hub_proto::types::InterfaceJson,
+    ) -> Result<Self, Self::Error> {
+        let interface_str = String::from_utf8_lossy(&interface.0);
+
+        crate::Interface::from_str(interface_str.as_ref()).map_err(Self::Error::Interface)
+    }
+}
+
+// FIXME this is probably useless ?
+impl TryFrom<AstarteDeviceDataEvent> for astarte_message_hub_proto::AstarteMessage {
+    type Error = MessageHubProtoError;
+
+    fn try_from(value: AstarteDeviceDataEvent) -> Result<Self, Self::Error> {
+        use astarte_message_hub_proto::astarte_message::Payload;
```

Comment:
It seems you are using `Payload` several times. Why don't you import it directly on the top of this file?

Reply:
There are many similar names between the structs defined in the sdk and the one generated from the protobuf files. I think it has been done to make it immediately clear to which package the type is referring to and I did not change it, if you think it would be better it can be done, for me it works either way.

Reply:
To me is also fine. Btw, you could think of renaming `Payload` so that there is no confusion (e.g., `ProtoPayload`, `MsgHubPayload` or similar)

Reply:
I did this in the end since I liked it better.